lib/Dialect/SPIRV/Transforms/DecorateWithLayoutInfoSPIRVPass.cpp

```diff
@@ -0,0 +1,329 @@
+//===- DecorateWithLayoutInfoSPIRVPass.cpp - Decorate with layout info ----===//
+//
+// Copyright 2019 The MLIR Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+//
+// This file implements a pass to decorate SPIR-V types with layout info.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/SPIRV/Passes.h"
+#include "mlir/Dialect/SPIRV/SPIRVDialect.h"
+#include "mlir/Dialect/SPIRV/SPIRVOps.h"
+#include "mlir/Transforms/DialectConversion.h"
+
+using namespace mlir;
+
+/// According to the SPIR-V spec:
+/// "There are different alignment requirements depending on the specific
+/// resources and on the features enabled on the device.
+/// There are 3 types of alignment: scalar, base, extended.
+///
+/// Note: Even if scalar alignment is supported, it is generally more
+/// performant to use the base alignment.
+///
+/// The memory layout must obey the following rules:
+/// 1.The Offset decoration of any member must be a multiple of its alignment.
+/// 2.Any ArrayStride or MatrixStride decoration must be a multiple of the
+/// alignment of the array or matrix as defined above.
+/// 3.The ArrayStride, MatrixStride, and Offset decorations must be large
+/// enough to hold the size of the objects they affect (that is, specifying
+/// overlap is invalid).
+class SPIRVLayoutInfoUtils {
+public:
+  using Alignment = uint64_t;
+
+  /// Return a new spirv::StructType with layout info, increment a base size by
+  /// proccessed type size in bytes and populate an alignment.
+  static Type decorateType(spirv::StructType structType,
```

Comment:
So these functions, with all the match calculations inside, can be quite hard to understand. What about explaining the meaning of `baseSize` and `baseAlignment` in comments?